# load packages

import pandas as pd
import os
from snakemake.utils import validate
from snakemake.utils import min_version

min_version("9.5.1")

# configuration
configfile: "../config/config.yaml"

validate(config, "../config/config.schema.yaml")

# container version of workflow
#containerized: "docker://milesroberts/poolseq-kmers"


# load samples
#parameters = pd.read_table(
#    "../config/parameters.tsv", sep="\t", dtype={"ID": str}
#).set_index(["ID"], drop=False)

#validate(parameters, "../config/parameters.schema.yaml")

# Define rules too fast to bother with cluster submission
localrules: all, batch_per_sample

rule all:
    input:
        expand("samtools_stats/{ID}_{linref}.txt", ID = config["samples"], linref = config["linrefs"]),
        expand("bcftools_stats_results/{linref}.txt", linref = config["linrefs"]),
        expand("mk_test_{linref}/mk.tsv", linref = config["linrefs"]),
        expand("salmon_quantmerge_results/{linref}.txt", linref = config["linrefs"]),
        expand("pixy_results/{linref}_{stat}.txt", linref = config["linrefs"], stat = ["pi", "watterson_theta", "tajima_d", "dxy", "fst"]),
        "kmer_distances.txt",
        "multiqc_report.html"

rule batch_per_sample:
    input:
        expand("bcftools_linref_results/{ID}_{ref}_{split}.vcf.gz", ID = config["samples"], ref = config["linrefs"], split = range(10, 10 + config["splits"])),
        expand("cbf_results/{ID}.txt", ID = config["samples"]),
        expand("salmon_quant_results/{ref}_{rna}/quant.sf", ref = config["linrefs"], rna = config["rna"])

## data download and QC rules
include: "rules/datasets.smk"
include: "rules/sra.smk"
include: "rules/fastp.smk"
include: "rules/multiqc.smk"

## linear alignment rules
include: "rules/bwa_index.smk"
include: "rules/bwa_mem.smk"
include: "rules/mark_dup.smk"

## genotype calling rules
include: "rules/bcftools_call.smk"
include: "rules/bcftools_merge.smk"
include: "rules/bcftools_filter.smk"

## k-mer counting rules
include: "rules/datasets.smk"
include: "rules/kmc_contam_db.smk"
include: "rules/kmc_rm_contam.smk"
include: "rules/kmc.smk"
include: "rules/counting_bloom_filter.smk"
include: "rules/cbind.smk"
include: "rules/kmer_distances.smk"

## pangenome alignment rules
include: "rules/vg_index.smk"
include: "rules/vg_giraffe.smk"
include: "rules/vg_filter.smk"
include: "rules/vg_surject.smk"
include: "rules/vg_paths.smk"

## gene expression quantification
include: "rules/salmon_index.smk"
include: "rules/salmon_quantmerge.smk"
include: "rules/salmon_quant.smk"

## population genetics summary rules
include: "rules/mk_test.smk"
include: "rules/degenotate.smk"
include: "rules/split_sites.smk"
include: "rules/pixy.smk"
