# vg-kmer-ref-snakes

A snakemake workflow to analyze short reads with a either pangenome reference, linear reference, and/or no reference

# To-do

- [x] add mitochondrial and chloroplast genomes as contaminants

- [ ] figure out vg surject

- [ ] add genotype calling

- [ ] add kmer counts

- [ ] remove k-mers from cds sequences?

- [ ] counting bloom filter
